### PR TITLE
Added multiple_policies_for_same_node.go e2e test

### PR DESCRIPTION
**Pull request by eranco74**
Added multiple_policies_for_same_node.go e2e test

```release-note
NONE
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -49,6 +49,7 @@ resources = deploy/namespace.yaml deploy/service_account.yaml deploy/role.yaml d
 all: check handler
 
 check: format vet whitespace-check
+	hack/check-go-version.sh
```
Comment (dankenigsberg) [inline Makefile]: seems like a very good idea, but unrelated to the purpose of this PR. Please split to another PR with a proper title and purpose.

File: test/e2e/multiple_policies_for_same_node.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		})
+		It("should have the IP and vlan interface configured", func() {
+			for _, node := range nodes {
+				ipAddressForNodeInterfaceEventually(node, *firstSecondaryNic).Should(Equal(ipAddress))
+				interfacesNameForNodeEventually(node).Should(ContainElement(fmt.Sprintf(`%s.%s`, *firstSecondaryNic, vlanId)))
+				vlanForNodeInterfaceEventually(node, fmt.Sprintf(`%s.%s`, *firstSecondaryNic, vlanId)).Should(Equal(vlanId))
+			}
+		})
+	})
+})
```
Comment (dankenigsberg) [inline test/e2e/multiple_policies_for_same_node.go]: is the rename
>  test/e2e/simple_bridge_and_bond.go → test/e2e/simple_bridge_and_bond_test.go 
related to this PR? Is it intentional (you do not add _test to your new file)

File: test/e2e/utils.go [hunk 3]
```diff
@@ -493,7 +505,12 @@ func dhcpFlag(node string, name string) bool {
 	return gjson.ParseBytes(currentStateJSON(node)).Get(path).Bool()
 }
 
-func ipv4Address(node string, name string) string {
-	path := fmt.Sprintf("interfaces.#(name==\"%s\").ipv4.address.0.ip", name)
+func ipv4Address(node string, iface string) string {
```
Comment (dankenigsberg) [inline test/e2e/utils.go]: reasonable, but unrelated.

File: test/e2e/multiple_policies_for_same_node.go (hunk 2 above)
Comment (eranco74) [reply test/e2e/multiple_policies_for_same_node.go]: I made 4 commites.
Only one is related to the test.
I'll split the PR....

Comment (dankenigsberg) [reply test/e2e/multiple_policies_for_same_node.go]: urghh, I prefer gerrit review process to github's every day. If you think that these commits are related you may bulk them together, but please explain this in the PR leading message.

File: test/e2e/multiple_policies_for_same_node.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
+package e2e
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: Please see https://github.com/nmstate/kubernetes-nmstate/pull/351, add _test suffix to the file name

File: test/e2e/multiple_policies_for_same_node.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	nmstatev1alpha1 "github.com/nmstate/kubernetes-nmstate/pkg/apis/nmstate/v1alpha1"
+)
+
+func infUpWithStaticIP(infName string, ipAddress string) nmstatev1alpha1.State {
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: We use `iface`

File: test/e2e/multiple_policies_for_same_node.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        - ip: %s
+          prefix-length: 24
+        dhcp: false
+        enabled: true
+      ipv6:
+        enabled: false
+`, infName, ipAddress))
+}
+
+func infUpWithVlan(iface string, vlanId string) nmstatev1alpha1.State {
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: ifaceWithVlanUp

File: test/e2e/multiple_policies_for_same_node.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        enabled: false
+`, iface, ipAddress))
+}
+
+var _ = Describe("NodeNetworkState", func() {
+	Context("with multiple policies configured", func() {
+		var (
+			StaticIpPolicy = "static-ip-policy"
+			VlanPolicy     = "vlan-policy"
+			ipAddress      = string("10.244.0.1")
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: no need for `string`

File: test/e2e/multiple_policies_for_same_node.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+			vlanId         = string("102")
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: ditto

File: test/e2e/multiple_policies_for_same_node.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+		)
+
+		BeforeEach(func() {
+			setDesiredStateWithPolicy(StaticIpPolicy, infUpWithStaticIP(*firstSecondaryNic, ipAddress))
+			setDesiredStateWithPolicy(VlanPolicy, infUpWithVlan(*firstSecondaryNic, vlanId))
+			waitForAvailablePolicy(StaticIpPolicy)
+			waitForAvailablePolicy(VlanPolicy)
+		})
+		AfterEach(func() {
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: please insert an empty line above

File: test/e2e/multiple_policies_for_same_node.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+			setDesiredStateWithPolicy(StaticIpPolicy, infStaticIPAbsent(*firstSecondaryNic, ipAddress))
+			setDesiredStateWithPolicy(VlanPolicy, infVlanAbsent(*firstSecondaryNic, vlanId))
+			deletePolicy(StaticIpPolicy)
+			deletePolicy(VlanPolicy)
+			resetDesiredStateForNodes()
+		})
+		It("should have the IP and vlan interface configured", func() {
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: ditto

**Comment by phoracek** [comment]
@qinqon seems like the webhook is failing, please see https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/348/pull-kubernetes-nmstate-e2e-k8s/1221720547679801344.

File: test/e2e/multiple_policies_for_same_node.go (hunk 4 above)
Comment (phoracek) [reply test/e2e/multiple_policies_for_same_node.go]: This was not resolved.

File: test/e2e/multiple_policies_for_same_node.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        enabled: true
+      ipv6:
+        enabled: false
+`, iface, ipAddress))
+}
+
+var _ = Describe("NodeNetworkState", func() {
+	Context("with multiple policies configured", func() {
+		var (
+			StaticIpPolicy = "static-ip-policy"
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: This does not need to be a public variable.

File: test/e2e/multiple_policies_for_same_node.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+			VlanPolicy     = "vlan-policy"
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: Ditto

File: test/e2e/multiple_policies_for_same_node.go [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	nmstatev1alpha1 "github.com/nmstate/kubernetes-nmstate/pkg/apis/nmstate/v1alpha1"
+)
+
+func ifaceUpWithStaticIP(iface string, ipAddress string) nmstatev1alpha1.State {
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node.go]: Could you make all these helpers into local functions of the `Context` they are needed in? We have quite a lot of these and they are shared across all the tests. Either make them private

```
ifaceUpWithStaticIP = func(...) ...
```

Or move them to `test/e2e/states.go`

File: test/e2e/multiple_policies_for_same_node_test.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+		BeforeEach(func() {
+			setDesiredStateWithPolicy(staticIpPolicy, ifaceUpWithStaticIP(*firstSecondaryNic, ipAddress))
+			setDesiredStateWithPolicy(vlanPolicy, ifaceUpWithVlanUp(*firstSecondaryNic, vlanId))
+			waitForAvailablePolicy(staticIpPolicy)
+			waitForAvailablePolicy(vlanPolicy)
+		})
+
+		AfterEach(func() {
+			setDesiredStateWithPolicy(staticIpPolicy, ifaceStaticIPAbsent(*firstSecondaryNic, ipAddress))
+			setDesiredStateWithPolicy(vlanPolicy, vlanAbsent(*firstSecondaryNic, vlanId))
```
Comment (qinqon) [inline test/e2e/multiple_policies_for_same_node_test.go]: You have to wait for this too with waitForAvailablePolicy

File: test/e2e/multiple_policies_for_same_node_test.go [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+		var (
+			staticIpPolicy = "static-ip-policy"
+			vlanPolicy     = "vlan-policy"
+			ipAddress      = "10.244.0.1"
+			vlanId         = "102"
+		)
+
+		BeforeEach(func() {
+			setDesiredStateWithPolicy(staticIpPolicy, ifaceUpWithStaticIP(*firstSecondaryNic, ipAddress))
+			setDesiredStateWithPolicy(vlanPolicy, ifaceUpWithVlanUp(*firstSecondaryNic, vlanId))
```
Comment (qinqon) [inline test/e2e/multiple_policies_for_same_node_test.go]: I would suggest to use secondSecondaryNic so we don't interface with the other policy, if this is what we want.

File: test/e2e/multiple_policies_for_same_node_test.go [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+			waitForAvailablePolicy(staticIpPolicy)
+			waitForAvailablePolicy(vlanPolicy)
+		})
+
+		AfterEach(func() {
+			setDesiredStateWithPolicy(staticIpPolicy, ifaceStaticIPAbsent(*firstSecondaryNic, ipAddress))
```
Comment (qinqon) [inline test/e2e/multiple_policies_for_same_node_test.go]: Don't do absent on ethernet nic (you cann't do that) just remove ip configuration

File: test/e2e/states.go [hunk 17]
```diff
@@ -89,3 +89,57 @@ func ovsbBrWithInternalInterface(bridgeName string) nmstatev1alpha1.State {
...
+    - name: %s.%s
+      type: vlan
+      state: absent
+      vlan:
+        base-iface: %s
+        id: %s
+`, iface, vlanId, iface, vlanId))
+}
+
+func ifaceStaticIPAbsent(iface string, ipAddress string) nmstatev1alpha1.State {
```
Comment (qinqon) [inline test/e2e/states.go]: ditto

File: test/e2e/multiple_policies_for_same_node_test.go (hunk 15 above)
Comment (eranco74) [reply test/e2e/multiple_policies_for_same_node_test.go]: I think the test is better if both policies use the same interface.
Since it seems to work I believe we should keep it.

**Comment by eranco74** [comment]
@phoracek please review

File: test/e2e/multiple_policies_for_same_node_test.go [hunk 18]
```diff
@@ -0,0 +1,44 @@
+package e2e
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("NodeNetworkState:MultiplePolicies", func() {
```
Comment (phoracek) [inline test/e2e/multiple_policies_for_same_node_test.go]: `:MultiplePolicies` is redundant with the following `Context`

**Comment by phoracek** [comment]
/hold

Until we resolve the failing CI

**Comment by qinqon** [comment]
/hold cancel

ci is working fine

**Comment by eranco74** [comment]
@qinqon the ci is not stable after all.
I'll ask the bot to retest.
Also @tsorya is adding the NetworkManager logs so it will be easier to find the root cause of the failure

/retest
